n3fit/src/evolven3fit/utils.py

```diff
@@ -0,0 +1,106 @@
+from scipy.interpolate import interp1d
+import numpy as np
+import math
+import pathlib
+import yaml
+import shutil
+
+
+class LhapdfLike:
+    pids_dict = {
+        -6: "TBAR",
+        -5: "BBAR",
+        -4: "CBAR",
+        -3: "SBAR",
+        -2: "UBAR",
+        -1: "DBAR",
+        21: "GLUON",
+        1: "D",
+        2: "U",
+        3: "S",
+        4: "C",
+        5: "B",
+        6: "T",
+        22: "PHT",
+    }
+    pids_order = [
+        "TBAR",
+        "BBAR",
+        "CBAR",
+        "SBAR",
+        "UBAR",
+        "DBAR",
+        "GLUON",
+        "D",
+        "U",
+        "S",
+        "C",
+        "B",
+        "T",
+        "PHT",
+    ]
+
+    def __init__(self, pdf_grid, q20, x_grid):
+        self.pdf_grid = pdf_grid
+        self.q20 = q20
+        self.x_grid = x_grid
+        self.funcs = [
+            interp1d(self.x_grid, self.pdf_grid[pid])
+            for pid in range(len(self.pids_order))
+        ]
+
+    def xfxQ2(self, pid, x, q2):
+        if not math.isclose(q2, self.q20, rel_tol=1e-6):
+            raise ValueError("The q2 requested is not the fitting scale of this pdf")
+        return x * (self.funcs[self.pids_order.index(self.pids_dict[pid])](x))
+
+    def hasFlavor(self, pid):
+        if pid in self.pids_dict.keys():
+            return True
+        return False
+
+
+def read_runcard(usr_path):
+    """
+    reads the runcard and returns the relevant information for evolven3fit
+    """
+    runcard_path = usr_path / "filter.yml"
+    with runcard_path.open() as fp:
+        data = yaml.safe_load(fp)
+    return data
+
+
+def generate_q2grid(Q0, Qfin):
+    """
+    Generate the q2grid used in the final evolved pdfs (Temporary solution)
+    """
+    grid = np.geomspace(Q0 ** 2, Qfin ** 2, num=100)
+    return grid
+
+
+def generate_x_grid():
+    """
+    Generate the xgrid used for the eko
+    """
+    grid = np.geomspace(1e-09, 1.0, num=196)
```

Comment:
Here, most likely, we want to extract the xgrid from the theory (it is encoded in FkTables, if needed we can make it more comfortable to access)

Reply:
Actually, this is a bit tricky (I encountered this already some time ago, but back then we were far from the end), because so far official NNPDF grids on LHAPDF are having this range (I'm not sure about the actual mathematical form, but about the range) - however, the default PineAPPL grid is only ranging down to 2e-7. This means there would be a genuine extrapolation region that has nothing to do with data non-availability, but only with numerical interpolation. Discussing with @enocera we agreed that this does not make sense and we should use instead the FK range.

Reply:
So, I understand you're confirming that we should use a single xgrid per theory, isn't it?

Reply:
yes (but this has to be communicated to the collaboration since this is affecting our one and only deliverable)

Reply:
Yes both the q2 and the x grid are temporary.